### ISSUE TITLE
Output Resolution / Definition Issue

### ISSUE THREAD

**Issue by amoebatron** [bytedance/LatentSync#67]
When I use LatentSync with a 1920 x 1080 input video, after processing, it gives me an output video that is also 1920 x 1080, however the quality of the video is noticeably more blurred than the original suggesting some kind of video quality downgrade. This is true for the entire output frame, not just the region around the mouth.

Is there any information as to why this happens? Or any workaround solution?

**Comment by chunyu-li** [comment]
can you paste the example video here? thanks

**Comment by amoebatron** [comment]
Sure. Here is the input video, and also the processed output video. As you should see, the quality of the output video has degraded during lipsync processing.

https://github.com/user-attachments/assets/48a78cf2-5e2f-4d1a-8653-08c08cd9bc9d


https://github.com/user-attachments/assets/d2fde2c8-c17b-4ec5-869b-f2d222dc375c

**Comment by chunyu-li** [comment]
Got it.
By the way, I am curious how do you generate the input video? It looks amazing.

**Comment by amoebatron** [comment]
Thanks. So for the input video I'm using a combination of Flux and Kling.
Flux to generate a single image, and then Kling to animate it.

**Comment by sultanchamberlain** [comment]
I am having a similar issue. The quality of my output video is significant poorer than my input video. Is this from the decoding the latent predictions or is this from preprocess/postprocessing of videos
Original VIdeo
https://github.com/user-attachments/assets/0c590e4c-7836-40de-bbd7-cea54d8f686e
LatentSync Output
https://github.com/user-attachments/assets/2b616272-bd9f-4717-a307-e6d8e16a5b5a

**Comment by sultanchamberlain** [comment]
I am having a similar issue. The quality of my output video is significant poorer than my input video. Is this from the decoding the latent predictions or is this from preprocess/postprocessing of videos
Original VIdeo

https://github.com/user-attachments/assets/49b71af8-5e5e-4dd8-b429-063ea32b9ef1

LatentSync Output

https://github.com/user-attachments/assets/062c3637-9b96-4de3-bc47-fef08bb36775

**Comment by icerao** [comment]
因为现在作者的代码，人脸跟整个画面帧融合是使用了整个人脸跟画面进行融合，可优化成例如生成的脸部画面的下面的部分去融合，这样眼睛这基本上就跟之前清晰度差不多了。

**Comment by sultanchamberlain** [comment]
Okay that is a good idea, however that only fixes the eyes. The actually prediction of the the mouth region is blurry, thus it means it is cause from the predicted latent and decoded image

**Comment by sanebg** [comment]
This is a very critical issue I also have. No matter what is my input the output is always messy sh*t and super rekt. If I mask only the mouth region that makes it even more unnatural, because everything else is sharp but the mouth is low res/blurry. We need to fix it ASAP; otherwise, this is not usable in real-world scenarios.

**Comment by sultanchamberlain** [comment]
I could be wrong, but the blurriness may stem from several sources:
	1.	The authors trained their model using the preprocessed VoxCeleb2 dataset as provided by [Oxford University](https://www.robots.ox.ac.uk/~vgg/data/voxceleb/vox2.html).
	2.	Reconstructed images from autoencoders are often blurry. In contrast, models that completely disentangle facial features, like those described in papers such as [EMOPortraits](https://github.com/neeek2303/EMOPortraits) and [VASA](https://www.microsoft.com/en-us/research/project/vasa-1/), tend to produce high-resolution images. For instance, Facebook’s [MARLIN](https://github.com/ControlNet/MARLIN) facial encoder also generates blurry outputs.

The preprocessed VoxCeleb2 dataset indiscriminately cropped faces from YouTube videos without filtering for very small bounding boxes. As a result, most of the videos are blurry, failing to capture finer details, such as teeth, but managing reasonably well to depict lips, jawlines, and general facial structure. To address this, I have re-cropped the 1.1 million VoxCeleb2 YouTube videos and selected only crops that are larger than 256x256 pixels. However, I lack the computational resources—such as 8 H100 GPUs—to train on this improved dataset. If anyone has access to sufficient compute resources, I’d like to collaborate using this dataset.

Another potential issue could lie with autoencoders themselves, which may not perform as well as methods that completely disentangle facial features. This is just a theory, but it aligns with my observations of other autoencoders like MARLIN, which also produces blurry outputs. To steelman the argument, MARLIN works at a resolution of 224x224, whereas VASA operates at 512x512. However, VASA’s authors have not released their code, so their demo results could be cherry-picked.

In conclusion, I believe we should train on a true 256x256 VoxCeleb2 dataset. The dataset provided by Oxford essentially reflects the limitations seen in LatentSync’s output. By correctly filtering for only 256x256 face crops, I have created a more robust dataset, and I think training with it will yield promising results.

**Comment by Souvic** [comment]
> I could be wrong, but the blurriness may stem from several sources: 1. The authors trained their model using the preprocessed VoxCeleb2 dataset as provided by [Oxford University](https://www.robots.ox.ac.uk/~vgg/data/voxceleb/vox2.html). 2. Reconstructed images from autoencoders are often blurry. In contrast, models that completely disentangle facial features, like those described in papers such as [EMOPortraits](https://github.com/neeek2303/EMOPortraits) and [VASA](https://www.microsoft.com/en-us/research/project/vasa-1/), tend to produce high-resolution images. Other facial autoencoder like Facebook’s [MARLIN](https://github.com/ControlNet/MARLIN) also generates blurry outputs.
> 
> The preprocessed VoxCeleb2 dataset indiscriminately cropped faces from YouTube videos without filtering for very small bounding boxes. As a result, most of the videos are blurry, failing to capture finer details, such as teeth, but managing reasonably well to depict lips, jawlines, and general facial structure. To address this, I have re-cropped the 1.1 million VoxCeleb2 YouTube videos and selected only crops that are larger than 256x256 pixels. However, I lack the computational resources—such as 8 H100 GPUs—to train on this improved dataset. If anyone has access to sufficient compute resources, I’d like to collaborate using this dataset.
> 
> Another potential issue could lie with autoencoders themselves, which may not perform as well as methods that completely disentangle facial features. This is just a theory, but it aligns with my observations of other autoencoders like MARLIN, which also produces blurry outputs. To steelman the argument, MARLIN works at a resolution of 224x224, whereas VASA operates at 512x512. In addition, VASA’s authors have not released their code, so their demo results could be cherry-picked.
> 
> In conclusion, I believe we should train on a true 256x256 VoxCeleb2 dataset. The dataset provided by Oxford essentially reflects the limitations seen in LatentSync’s output. By correctly filtering for only 256x256 face crops, I have created a more robust dataset, and I think training with it will yield promising results.

That's a series of very good observations. 
Can't we just fine-tune it though on a new dataset?
Before fine-tuning/training, I guess, we will need to check the output of the frozen decoder first to understand if indeed VAE encoder is culprit.

**Comment by markrmiller** [comment]
I do it at 720p and upscale it after. Works better. Topaz is good, but there are lots of free video upscale models.

There can also be issues depending on the movement of the head. Seems to be from detection box and landmark jitter. I changed it to do a double extraction pass like deepfacelab does and changed the smoothing to a kalman filter and got better results for the vids that were giving me larger issues.

**Comment by bbecausereasonss** [comment]
I'm getting terrible results even with low resolution videos, the quality is always reduced substantially.

**Comment by MarisKay** [comment]
I understand that there is some blurriness within the face region, no problem, i run Reactor on top of it and it gets the face back, BUT - why the rest of the area gets pixelated? Like why do you even touch it when you have to work only on face portion? Thanks

**Comment by markrmiller** [comment]
The area around the face or the whole image? If you mean the area around the face, the mask that is used is larger than the face, I'm assuming so that it's sure to fit any face shape.

**Comment by MarisKay** [comment]
Yeah, but unfortunately Sonic isjust img2video, that is kind of sad.

**Comment by MarisKay** [comment]
that could a potential partial workaround, for sure, though , what to do with lips? they are so out of context due to low quality output

**Comment by Momotyust** [comment]
your guys need GFPGAN or GPEN, after the output video of this repo

**Comment by MarisKay** [comment]
GFPGAN or any other face restoration model makes your person speak through teeth, mouth barely opens, doesnt look real.

**Comment by ross-Hr** [comment]
> I could be wrong, but the blurriness may stem from several sources: 1. The authors trained their model using the preprocessed VoxCeleb2 dataset as provided by [Oxford University](https://www.robots.ox.ac.uk/~vgg/data/voxceleb/vox2.html). 2. Reconstructed images from autoencoders are often blurry. In contrast, models that completely disentangle facial features, like those described in papers such as [EMOPortraits](https://github.com/neeek2303/EMOPortraits) and [VASA](https://www.microsoft.com/en-us/research/project/vasa-1/), tend to produce high-resolution images. Other facial autoencoder like Facebook’s [MARLIN](https://github.com/ControlNet/MARLIN) also generates blurry outputs.
> 
> The preprocessed VoxCeleb2 dataset indiscriminately cropped faces from YouTube videos without filtering for very small bounding boxes. As a result, most of the videos are blurry, failing to capture finer details, such as teeth, but managing reasonably well to depict lips, jawlines, and general facial structure. To address this, I have re-cropped the 1.1 million VoxCeleb2 YouTube videos and selected only crops that are larger than 256x256 pixels. However, I lack the computational resources—such as 8 H100 GPUs—to train on this improved dataset. If anyone has access to sufficient compute resources, I’d like to collaborate using this dataset.
> 
> Another potential issue could lie with autoencoders themselves, which may not perform as well as methods that completely disentangle facial features. This is just a theory, but it aligns with my observations of other autoencoders like MARLIN, which also produces blurry outputs. To steelman the argument, MARLIN works at a resolution of 224x224, whereas VASA operates at 512x512. In addition, VASA’s authors have not released their code, so their demo results could be cherry-picked.
> 
> In conclusion, I believe we should train on a true 256x256 VoxCeleb2 dataset. The dataset provided by Oxford essentially reflects the limitations seen in LatentSync’s output. By correctly filtering for only 256x256 face crops, I have created a more robust dataset, and I think training with it will yield promising results.

The gay has vRAMs:  https://github.com/bytedance/LatentSync/issues/151#issue-2899564267

**Comment by movieproducer** [comment]
If you want to fix the issue, in the \ComfyUI\custom_nodes\ComfyUI-LatentSyncWrapper directory open the file nodes.py.

Around line 495 replace these two lines:

```
import torchvision.io as io
io.write_video(temp_video_path, frames_cpu, fps=25, video_codec='h264')
```

with the following code:

             ```
   import torchvision.io as io  # Needed later when reading video
                import ffmpeg
                from PIL import Image

                # Create a directory to store temporary frame images
                frame_dir = os.path.join(temp_dir, "frames")
                os.makedirs(frame_dir, exist_ok=True)

                # Save each frame as PNG
                for i, frame in enumerate(frames_cpu):
                    img = Image.fromarray(frame.numpy())
                    frame_path = os.path.join(frame_dir, f"{i:04d}.png")
                    img.save(frame_path)

                # Use ffmpeg to encode the frames into a high-quality MP4 video
                try:
                    (
                        ffmpeg
                        .input(os.path.join(frame_dir, '%04d.png'), framerate=25)
                        .output(temp_video_path, vcodec='libx264', crf=18, pix_fmt='yuv420p', preset='slow')
                        .run(quiet=True, overwrite_output=True)
                    )
                except ffmpeg.Error as e:
                    print("FFmpeg error:", e.stderr.decode() if e.stderr else str(e))
                    raise RuntimeError("Failed to encode video with ffmpeg")

                # Clean up the temporary frame images
                shutil.rmtree(frame_dir, ignore_errors=True)

```


I tested it and it works perfect zero video degradation. you can control the quality of the video on this line:

`.output(temp_video_path, vcodec='libx264', crf=18, pix_fmt='yuv420p', preset='slow')`

It probably would not be hard to incorporate these setting so they are adjustable int he node itself, I may get around to trying it.

**Comment by movieproducer** [comment]
To be clear I used the a version of Latent Sync found here: https://github.com/ShmuelRonen/ComfyUI-LatentSyncWrapper

This is a Comfyui custom node and it is not the standard version you will find int he Comfyui node manager. 

I cloned the repo and renamed the folder so I can run both node version at the same time.

I am running Comfyui Local with an RTX4090.

When I make the code changes stated in my previous post it works and the video quality is much better than without the mod.

I use ffmpeg settings for best quality

 .output(temp_video_path, vcodec='libx264', crf=0, pix_fmt='yuv444p', preset='veryslow')
